sqlmesh/core/plan/definition.py

```diff
@@ -451,30 +452,48 @@ def has_unmodified_unpromoted(self) -> bool:
-            if table in self.context_diff.snapshots:
-                downstream.append(table)
-
-            snapshots = self.context_diff.snapshots
-            downstream = [
-                d for d in downstream if not snapshots[d].is_symbolic and not snapshots[d].is_seed
-            ]
-
-            if not self.is_dev:
-                models_with_disabled_restatement = [
-                    f"'{d}'" for d in downstream if snapshots[d].model.disable_restatement
-                ]
-                if models_with_disabled_restatement:
-                    raise PlanError(
-                        f"Restatement is disabled for models: {', '.join(models_with_disabled_restatement)}."
-                    )
-
-            if not downstream:
+        def is_restateable_snapshot(snapshot: Snapshot) -> bool:
+            if not self.is_dev and snapshot.model.disable_restatement:
+                logger.debug("Restatement is disabled for model '%s'.", snapshot.name)
+                return False
+            return not snapshot.is_symbolic and not snapshot.is_seed
+
+        dummy_interval = (sys.maxsize, -sys.maxsize)
+        snapshot_mapping = self.context_diff.snapshots
+        restatements = {}
+        # Add restate snapshots and their downstream snapshots
+        for snapshot_name in restate_models:
+            if snapshot_name not in snapshot_mapping or not is_restateable_snapshot(
+                snapshot_mapping[snapshot_name]
+            ):
                 raise PlanError(
-                    f"Cannot restate from '{table}'. Either such model doesn't exist, no other materialized model references it, or restatement was disabled for this model."
+                    f"Cannot restate from '{snapshot_name}'. Either such model doesn't exist, no other materialized model references it, or restatement was disabled for this model."
                 )
-            self._restatements.update(downstream)
+            restatements[snapshot_name] = dummy_interval
+            for downstream_snapshot_name in self._dag.downstream(snapshot_name):
+                if is_restateable_snapshot(snapshot_mapping[downstream_snapshot_name]):
+                    restatements[downstream_snapshot_name] = dummy_interval
+        # Get restatement intervals for all restated snapshots and make sure that if a snapshot expands it's
+        # restatement range that it's downstream dependencies all expand their restatement ranges as well.
+        for snapshot_name in self._dag:
+            if snapshot_name not in restatements:
+                continue
+            snapshot = snapshot_mapping[snapshot_name]
+            interval = snapshot.get_removal_interval(
+                self.start, self.end, self._execution_time, strict=False
+            )
+            upstream_snapshots = [
+                s
+                for s in self._dag.upstream(snapshot_name)
+                if is_restateable_snapshot(snapshot_mapping[s])
```

Comment:
Does it matter here? Didn't we already make sure that all snapshots are restateable?

Reply:
No at this point we have made sure that all restated snapshots are restateable and their downstream dependencies. Here we take a snapshot and then go upstream so we need to do this check again because this could include new snapshots that we haven't visited yet.

Reply:
I guess technically if it isn't restateable then it won't be in the restateable dict and therefore we can pull them in here and we will just get a miss on the following line when we check the dict.

Reply:
So I do think it can be removed.